simd/keccakf1600/f1600x.go

```diff
@@ -104,46 +102,54 @@ func (s *StateX2) Initialize() []uint64 {
 
 // Permute performs the four parallel Keccak-f[1600]s interleaved on the slice
 // returned from Initialize().
-func (s *StateX4) Permute() {
+// If turbo is true, applies 12-round variant instead of the usual 24.
+func (s *StateX4) Permute(turbo bool) {
 	if IsEnabledX4() {
-		permuteSIMDx4(s.a[s.offset:])
+		permuteSIMDx4(s.a[s.offset:], turbo)
 	} else {
-		permuteScalarX4(s.a[s.offset:]) // A slower generic implementation.
+		permuteScalarX4(s.a[s.offset:], turbo) // A slower generic implementation.
 	}
 }
 
 // Permute performs the two parallel Keccak-f[1600]s interleaved on the slice
 // returned from Initialize().
-func (s *StateX2) Permute() {
+// If turbo is true, applies 12-round variant instead of the usual 24.
+func (s *StateX2) Permute(turbo bool) {
 	if IsEnabledX2() {
-		permuteSIMDx2(s.a[s.offset:])
+		permuteSIMDx2(s.a[s.offset:], turbo)
 	} else {
-		permuteScalarX2(s.a[s.offset:]) // A slower generic implementation.
+		permuteScalarX2(s.a[s.offset:], turbo) // A slower generic implementation.
 	}
 }
 
-func permuteScalarX4(a []uint64) {
+func permuteScalarX4(a []uint64, turbo bool) {
 	var buf [25]uint64
 	for i := 0; i < 4; i++ {
 		for j := 0; j < 25; j++ {
 			buf[j] = a[4*j+i]
 		}
-		sha3.KeccakF1600(&buf)
+		sha3.KeccakF1600(&buf, turbo)
 		for j := 0; j < 25; j++ {
 			a[4*j+i] = buf[j]
 		}
 	}
 }
 
-func permuteScalarX2(a []uint64) {
+func permuteScalarX2(a []uint64, turbo bool) {
 	var buf [25]uint64
 	for i := 0; i < 2; i++ {
 		for j := 0; j < 25; j++ {
 			buf[j] = a[2*j+i]
 		}
-		sha3.KeccakF1600(&buf)
+		sha3.KeccakF1600(&buf, turbo)
 		for j := 0; j < 25; j++ {
 			a[2*j+i] = buf[j]
 		}
 	}
 }
+
+var enabledX2 bool
+
+func init() {
+	enabledX2 = runtime.GOARCH == "arm64" && runtime.GOOS == "darwin"
```

Comment:
does the code is OS-dependent? we only need to check whether the NEON/ASIMD instructions are supported right?

Reply:
No, but any Mac on ARM64 supports these extensions. We're just waiting for Go to add the feature flag, which they're refusing to do for the moment: https://github.com/golang/go/pull/46300